src/main/java/mate/academy/carsharingapp/service/impl/RentalServiceImpl.java

```diff
@@ -0,0 +1,94 @@
+    private final CarRepository carRepository;
+    private final NotificationService notificationService;
+
+    @Override
+    public RentalResponseDto createRental(Authentication authentication,
+                                          RentalRequestDto requestDto) {
+        Car car = carRepository.findById(requestDto.getCarId()).orElseThrow(
+                () -> new EntityNotFoundException("Car with id "
+                        + requestDto.getCarId() + " not found")
+        );
+        if (car.getInventory() <= 0) {
+            throw new IllegalStateException("Car with id " + car.getId() + " is not available");
+        }
+        Rental rental = new Rental();
+        User user = userRepository.findByEmail(authentication.getName()).orElseThrow(
+                () -> new EntityNotFoundException("User with email "
+                        + authentication.getName() + " not found")
+        );
+        rental.setRentalDate(LocalDate.now());
+        rental.setReturnDate(LocalDate.now().plusWeeks(1));
+        rental.setActualReturnDate(null);
+        rental.setUser(user);
+        rental.setCar(car);
+        car.setInventory(car.getInventory() - 1);
+        carRepository.save(car);
+        Rental saved = rentalRepository.save(rental);
+        notificationService.sendRentalCreationNotification(saved);
+        return rentalMapper.toDto(rental);
+    }
+
+    public List<RentalResponseDto> getRentalsByUserId(Long userId, Boolean isActive) {
+        return rentalRepository.getRentalsByUser_Id(userId).stream()
+                .filter(rental -> isActive == null
+                        || (isActive && rental.getActualReturnDate() == null)
+                        || (!isActive && rental.getActualReturnDate() != null))
+                .map(rentalMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public RentalResponseDto getSpecificRentalByUser_Id_And_Rental_Id(Long userId, Long rentalId) {
+        Rental rental = rentalRepository
+                .findByUser_IdAndId(userId, rentalId)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find rental by "
+                        + "user id " + userId + ", rental id " + rentalId));
+        return rentalMapper.toDto(rental);
+    }
+
+    @Override
+    public RentalResponseDto updateRentalReturnDate(Long rentalId, RentalUpdateDto updateDto) {
+        Rental rental = rentalRepository.findById(rentalId)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find rental by id "
+                        + rentalId));
+        if (rental.getActualReturnDate() != null) {
+            throw new IllegalStateException("Rental has already been returned");
+        }
+        rental.setActualReturnDate(updateDto.getReturnDate());
+        Car car = rental.getCar();
+        car.setInventory(car.getInventory() + 1);
+        carRepository.save(car);
+        Rental saved = rentalRepository.save(rental);
```

Comment:
saved - bad naming
you don't need this variable actually
```suggestion
        rentalRepository.save(rental);
```